[ORCA-323] Annotate submissions with locations of their output files

jaymedina opened the pull request:
### problem

The current model-to-data workflow does not update the submissions with annotations showing the synids of the output files generated for the given submission. It would be nice to have this feature because it makes it easier for Challenge organizers to quickly retrieve these output files instead of having to parse through the files themselves.

### solution

- [x] Update `update_folders.py` with a new piece of logic that also returns the synids of the output files so that they can be annotated in a subsequent task
- [x] Add a new task called `ANNOTATE_SUBMISSION_AFTER_UPDATE_FOLDERS` which will use the JSON generated from the `UPDATE_FOLDERS` step to annotate the submissions with the annotations of interest (`predictions_id`, the synid of the predictions file, and `docker_logs_id`, the synid of the Docker execution log)
- [x] Update documentation

Out of scope:
- [x] Fix concurrency issues with RUN_DOCKER -> UPDATE_FOLDERS, and status updates that were caused by working with `submission_ch` input value rather than the submission ID outputs from the processes.

### testing & preview

- [x] [Test](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/5BRjKvOCUJBmNo) with a valid predictions file
- [x] [Test](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/1n1IOjsr5ulwtK) with an `INVALID` predictions file
- [x] [Test](https://tower.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/pegs-challenge-project/watch/QtRO74FXKBpAl) with multiple submissions

Filtering the submission view for only submissions with a non-empty `docker_logs_id` (which implies a docker exec. log was created and uploaded) shows the 6 submissions I made in the 3 tests above:

<img width="785" alt="image" src="https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/assets/32107699/8033a09e-b2d3-4bc2-adc0-65e6f0c45f9a">

[Clicking each of them takes you to the correct output file for the correct submission](https://www.synapse.org/Synapse:syn53770151/tables/#query/eyJzcWwiOiJTRUxFQ1QgaWQsIG5hbWUsIHN0YXR1cywgZG9ja2VyX2xvZ3NfaWQsIHByZWRpY3Rpb25zX2lkLCBzY29yZTEsIHNjb3JlMiwgc2NvcmUzLCBzY29yZV9lcnJvcnMsIHNjb3JlX3N0YXR1cywgdmFsaWRhdGlvbl9lcnJvcnMsIHZhbGlkYXRpb25fc3RhdHVzIEZST00gc3luNTM3NzAxNTEgV0hFUkUgZG9ja2VyX2xvZ3NfaWQgSVMgTk9UIE5VTEw7IiwgImluY2x1ZGVFbnRpdHlFdGFnIjp0cnVlLCAibGltaXQiOjI1fQ==)

jaymedina committed:
Generate an output_annotation.json with the output file synid

jaymedina commented on modules/update_folders.nf:
```diff
@@ -11,7 +11,7 @@ process UPDATE_FOLDERS {
 
     output:
-    val "ready"
+    path "output_annotation.json"
```
This might be a problem in cases where an INVALID predictions file is being passed thru the pipeline. Investigate.

jaymedina replied:
Update: This is not a problem because the Docker execution log will always be created and uploaded even if the INVALID file is not, so the `output_annotation.json` will always be created.

jaymedina committed:
New task: ANNOTATE_SUBMISSION_AFTER_UPDATE_FOLDERS. Revamping params

jaymedina commented on bin/update_folders.py:
```diff
@@ -105,6 +106,13 @@ def update_folders(
+    output_annotation = {f"{folder_name}_id": file_synid}
+
+    with open("output_annotation.json", "w") as o:
+        o.write(json.dumps(output_annotation))
```
This overwrite `output_annotations.json` created for predictions, so we only get the `docker_logs_id` annotation. Change this functionality such that it _appends_ to the content already in `output_annotation.json` (even if empty).

jaymedina commented on workflows/MODEL_TO_DATA.nf:
```diff
@@ -63,13 +64,14 @@ workflow MODEL_TO_DATA {
-    ANNOTATE_SUBMISSION_AFTER_SCORE(SCORE.output)
+    ANNOTATE_SUBMISSION_AFTER_SCORE(submission_ch, SCORE.output.map { it[3] })
     if (params.send_email) {
         SEND_EMAIL(params.email_script, params.view_id, submission_ch, "AFTER", params.email_with_score, ANNOTATE_SUBMISSION_AFTER_SCORE.output)
```
The e-mail now has these annotations as part of the score annotations, which shouldn't be happening.

jaymedina committed:
Prevent predictions_id annotation from getting overwritten

jaymedina committed:
Typo

jaymedina committed:
Prevent these new annotations from showing up in the e-mail

jaymedina committed:
Working with tuple inputs instead for cleanliness

jaymedina committed:
Reverting update_submission_status inputs

jaymedina committed:
Syntax fix: declare output type for status

jaymedina committed:
Update documentation

jaymedina committed:
Add a type argument to Dict

jaymedina committed:
Print stmts added for logs

jaymedina committed:
Rename output_annotation.json with submission id

jaymedina committed:
Print what file is being stored for what submission

jaymedina committed:
Explicit outputs

jaymedina committed:
String interpolation in run_docker.nf output to specify subids

jaymedina committed:
No string interpolation

jaymedina committed:
Explicit output channel for RUN_DOCKER

jaymedina committed:
remove dsl1 syntax

jaymedina committed:
syntax

jaymedina committed:
Updating expected channels

jaymedina committed:
explicit output handling

jaymedina committed:
Fixing out of sync update submission status

jaymedina committed:
Fix concurrency issue in status updates

BWMac commented on workflows/MODEL_TO_DATA.nf:
```diff
@@ -53,24 +53,49 @@ include { VALIDATE } from '../modules/validate_model_to_data.nf'
+    ANNOTATE_SUBMISSION_AFTER_UPDATE_FOLDERS(UPDATE_FOLDERS.output)
+
+    // Phase 3: Validation of Docker submission results
+    validate_outputs = VALIDATE(run_docker_outputs, SYNAPSE_STAGE_GOLDSTANDARD.output, UPDATE_SUBMISSION_STATUS_AFTER_RUN.output, params.execute_validation)
```
Just a heads up that I think this will run while UPDATE_FOLDERS and ANNOTATE_SUBMISSION_AFTER_UPDATE_FOLDERS are running in case that is unintended.

jaymedina replied:
That's fine - As long as this runs after the submission status update, it won't mess anything up. The `VALIDATE` step is contingent on the `UPDATE_SUBMISSION_STATUS_AFTER_RUN` output which is what we want. Updating the annotation and uploading the files to Synapse are independent of validation 👍

BWMac commented on workflows/MODEL_TO_DATA.nf:
```diff
@@ -53,24 +53,49 @@ include { VALIDATE } from '../modules/validate_model_to_data.nf'
 
 workflow MODEL_TO_DATA {
+
+    // Phase 0: Each submission is evaluated in its own separate channel
```
I like the comments dividing the workflow logically up into phases. You might consider looking at turning all of these into formal subworkflows in the future.

BWMac commented on workflows/MODEL_TO_DATA.nf:
```diff
@@ -53,24 +53,49 @@ include { VALIDATE } from '../modules/validate_model_to_data.nf'
+    // Phase 2: Running the Docker submission (runs after Phase 1 data staging)
+    run_docker_outputs = RUN_DOCKER(submission_ch, params.container_timeout, params.poll_interval, SYNAPSE_STAGE_DATA.output, params.cpus, params.memory, params.log_max_size, CREATE_FOLDERS.output, UPDATE_SUBMISSION_STATUS_BEFORE_RUN.output)
+    //// Explicit output handling
+    run_docker_submission = run_docker_outputs.map { submission_id, predictions, logs -> submission_id }
+    run_docker_files = run_docker_outputs.map { submission_id, predictions, logs -> tuple(predictions, logs) }
```
What is your rationale for doing these `map` steps instead of just changing the `RUN_DOCKER` outputs to satisfy what is needed for both cases? (Same comment/question for the similar patterns below)

jaymedina replied:
The motivation was to improve readability in the workflow. When mapping the output directly within the process call (e.g. `RUN_DOCKER.map.{ it[0] }` ), it added an extra step of having to go into `run_docker.nf` to read the output getting indexed. This way helps minimize overhead when troubleshooting and understanding the pipeline.

BWMac commented on modules/update_folders.nf:
```diff
@@ -7,11 +7,10 @@ process UPDATE_FOLDERS {
 
     output:
-    val "ready"
+    tuple val(submission_id), path(predictions_file), val("status"), path("output_annotation*.json")
```
Rather than using a separate JSON file here for these annotations, would it work to just start the `results.json` file here so that we are using one common file for all annotations/results throughout the workflow?

jaymedina replied:
I think this would work, with the caveat that the annotations would be overwritten in the validation step: The PEGS challenge organizers' [validation script](https://github.com/Sage-Bionetworks-Challenges/pegs-evaluation/blob/aa69ecac04ea1a4dadf45990f989876940c472e9/validate.py#L140) for example writes into `results.json` with the assumption that it hasn't been made before. That line there will overwrite the contents in the `results.json` with just the validation results.

But by this point, the annotations have already been added to the submission on Synapse, so it wouldn't matter, except from a troubleshooting standpoint I guess?

At this point though, it would be the same as working with 2 separate files except the files have the same name. I think the value of working with 1 JSON is being able to pull out metadata and use them as input parameters, instead of using input params in the processes themselves, but I don't think we'd get this value until we do a full refactoring + loop in the CNB team for it.

BWMac replied:
Ah I see. I don't love using the same file name but overwriting it for the reason you outlined. I think we just keep an eye on this for the future as you say. Thanks!

BWMac commented on bin/update_folders.py:
```diff
@@ -105,6 +124,21 @@ def update_folders(
+    print(f"Synapse ID for {folder_name} is {file_synid}")
+
+    # Read existing data if file exists, otherwise `existing_annotation` is an empty dictionary
+    existing_annotation = load_data(output_annotation_filename)
```
Sorry if I missed this somewhere, but under what circumstances would this file already exist?

jaymedina replied:
`update_folders.py` gets executed twice in `update_folders.nf`: once for the predictions file (if it's not an `INVALID` file), and once for the docker log. This logic accounts for the execution for the docker log.

BWMac replied:
Ah right, thanks!

BryanFauble reviewed (approved):
Nothing additional beyond what Brad has already flagged! Great review! Will go ahead and pre-approve pending those fixes.

jaymedina committed:
Renaming input in annotate_submission for clarity

jaymedina committed:
use score_submission instead of score output map

jaymedina committed:
typo: validate_outputs.map shouldve been score_outputs.map

BWMac reviewed (approved):
Great work! 🔥

jaymedina commented:
I've made some changes since the last test run so [here is the new test run](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/15Ihz1jgAHng5F), and new screenshot:

<img width="788" alt="image" src="https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/assets/32107699/54a985ff-8904-4260-b58a-149d0dc4e394">

[Clicking through the annotation synids takes me to the correct output file for the correct submission](https://www.synapse.org/Synapse:syn53770151/tables/#query/eyJzcWwiOiJTRUxFQ1QgaWQsIG5hbWUsIHN0YXR1cywgZG9ja2VyX2xvZ3NfaWQsIHByZWRpY3Rpb25zX2lkLCBzY29yZTEsIHNjb3JlMiwgc2NvcmUzLCBzY29yZV9lcnJvcnMsIHNjb3JlX3N0YXR1cywgdmFsaWRhdGlvbl9lcnJvcnMsIHZhbGlkYXRpb25fc3RhdHVzIEZST00gc3luNTM3NzAxNTEgV0hFUkUgZG9ja2VyX2xvZ3NfaWQgSVMgTk9UIE5VTEw7IiwgImluY2x1ZGVFbnRpdHlFdGFnIjp0cnVlLCAibGltaXQiOjI1fQ==).

I'll go ahead and merge this now. Thanks!
